server/manager/src/main/java/org/apache/accumulo/manager/Manager.java

```diff
@@ -225,6 +230,40 @@ public boolean stillManager() {
     return getManagerState() != ManagerState.STOP;
   }
 
+  /**
+   * Retrieve the Fate object, blocking until it is ready. This could cause problems if Fate
+   * operations are attempted to be used prior to the Manager being ready for them. If these
+   * operations are triggered by a client side request from a tserver or client, it should be safe
+   * to wait to handle those until Fate is ready, but if it occurs during an upgrade, or some other
+   * time in the Manager before Fate is started, that may result in a deadlock and will need to be
+   * fixed.
+   *
+   * @return the Fate object, only after the fate components are running and ready
+   */
+  Fate<Manager> fate() {
+    try {
+      // block up to 30 seconds until it's ready; if it's still not ready, introduce some logging
+      if (!fateReadyLatch.await(30, TimeUnit.SECONDS)) {
+        String msgPrefix = "Unexpected use of fate in thread " + Thread.currentThread().getName()
+            + " at time " + System.currentTimeMillis();
+        // include stack trace so we know where it's coming from, in case we need to troubleshoot it
+        log.warn("{} blocked until fate starts", msgPrefix,
+            new IllegalStateException("Attempted fate action before manager finished starting up; "
+                + "if this doesn't make progress, please report it as a bug to the developers"));
+        int minutes = 0;
+        while (!fateReadyLatch.await(5, TimeUnit.MINUTES)) {
+          minutes += 5;
+          log.warn("{} still blocked after {} minutes; this is getting weird", msgPrefix, minutes);
+        }
+        log.debug("{} no longer blocked", msgPrefix);
+      }
```

Comment:
This will block the current thread but there are multiple threads trying to access the fate object. It seems like there is one thread listening per each RPC method. So I am not sure how useful all the logic and complexity added for the blocking is. I would prefer a less complex change where we don't do any thread blocking. I do like the improved logging message. Anything is better, is better then NPEs.

Reply:
Instead of blocking here in `Manager.fate()` we could throw an exception. When called from the `FateServiceHandler` that exception could be turned into a Thrift exception and returned back to the caller. Would that be a better solution?

Reply:
Maybe. We don't want to kill the thread though cause they were just started in the Manager as part of the startup.

Reply:
We were already throwing an exception... it was a NPE. I think it would have triggered a retry on the RPC loop, though. The whole point of this was to avoid the exception throwing.

Reply:
I'm not suggesting that we kill the thread, I'm suggesting that we throw an error in `Manager.fate()` when it's not ready, catch it in `FateServiceHandler`, and return a TException back to the originator of the call - likely the TabletServer.

Reply:
I think catching the exception, and throwing something the client code should handle differently is going to be a lot bigger of a change, involving updating the RPC method signatures, and adding client side code to handle it differently than the current loop.

I would prefer just giving it a bit of time, and waiting for the object to be ready, with some generous logging, as this PR is doing.

Reply:
Yeah, I agree with that too. I was just trying to understand what type of alternate solution would be acceptable.

Reply:
> I'm not suggesting that we kill the thread, I'm suggesting that we throw an error in `Manager.fate()` when it's not ready, catch it in `FateServiceHandler`, and return a TException back to the originator of the call - likely the TabletServer.

Since this seems to be a routine case, and things should normally make progress within a few seconds, I don't think it's worthwhile using Exception handling as flow control for the tablet server.